### DISCUSSION TITLE
Is it recommended to replace all the jotai call with jotai-scope's?

### DISCUSSION THREAD

**Discussion by yf-yang** [pmndrs/jotai#2204] [General]
@dai-shi Just when I am writing the question, I saw you merge https://github.com/jotaijs/jotai-scope/pull/5. I have one remaining question and need some guidance (maybe it should be shared to other jotai users).

Since jotai-scope has almost the same API with jotai, what is the performance overhead if I just replace all the jotai `useAtom` alike calls with jotai-scope's? Is it recommended?

**Comment by dai-shi** [comment]
Yes, it has an overhead, which increases with nesting ScopeProvider.
I tried best to make it efficient by caching atoms, but theoretically using the hooks from `jotai-scope` is less efficient.
For the actual performance difference, you want to measure it. (And, I'm interested in it, so you are highly encouraged.)

Meanwhile, I think one can try using `useAtom` for everything if they use `jotai-scope`. We wouldn't recommend it for normal Jotai users without requiring scope, though.

**Reply by yf-yang** [reply]
Emmm, I am willing to do so, but I've never written any React framework related benchmarks, do you have any suggestion or any good reference projects? Given a starting point, maybe we can do something similar.

**Reply by yf-yang** [reply]
@dai-shi check this out https://codesandbox.io/s/practical-moser-xjq5t2?file=/src/App.jsx
## Usage
After changing parameters, click `Mount Atoms` (I don't take mount overhead into account), then click Start measurement, and check the console.
Note that codesandbox is dev mode and enables React.StrictMode, so I use create next app to build one in my local desktop. It seems the result varies and does not show significant performance difference.

---

I am considering the following problems:
## Are the benchmark cases thorough and precisely reflect jotai's use cases?
I personally think maybe the benchmark does not show several scenarios:
  - Multiple primitive atoms
  - Multiple derived atom cascading each other
But are they crucial in comparison between two versions of useAtoms? I am not sure.
## Do we really care?
Maybe compared to other application code, the performance difference here is too slight. A supportive evidence is that standard deviation is several times larger than the mean, so I suspect other stuff the browser is doing makes more impact. The result is, the overall time does not give any clue, sometimes original useAtom is even slower than jotai-scope's.

Do you have any other suggestions or ideas?

**Reply by yf-yang** [reply]
After writing the analysis of https://github.com/jotaijs/jotai-scope/issues/8#issuecomment-1776665511, I add additional one layer of `ScopeProvider`, it shows that jotai-scope's useAtom is 2~3 times slower.
https://codesandbox.io/s/amazing-stonebraker-zxyv68?file=/src/App.jsx

This example illustrates the second scenario. The one above illustrates the first scenario.

**Comment by dai-shi** [comment]
#2205 is merged to follow the v0.3.0 change. It's just minimal. If you have suggestion on docs, please feel free to open PRs.

**Reply by yf-yang** [reply]
Yep, I just upgraded it. For the time being I know clearly what is the difference, but I can expect that the identical API name and similar use cases would soon obfuscate my collaborators. The easiest way to do so is replacing them completely, so you never worry about other developers' confusion. I have similar considerations when thinking about docs, it needs some thoughts.
